--- a/kernel/infra/file_wrapper.c
+++ b/kernel/infra/file_wrapper.c
@@ -85,6 +85,7 @@
     return orig->f_op->write_iter(iocb, iovi);
 }
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 1, 0)
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0)
 static int ksu_wrapper_iopoll(struct kiocb *kiocb, struct io_comp_batch *icb, unsigned int v)
 {
@@ -102,6 +103,7 @@
     return orig->f_op->iopoll(kiocb, spin);
 }
 #endif
+#endif /* iopoll >= 5.1 */
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(6, 6, 0)
 static int ksu_wrapper_iterate(struct file *fp, struct dir_context *dc)
@@ -302,6 +304,7 @@
     struct file *orig = data->orig;
     return orig->f_op->copy_file_range(file_in, pos_in, orig, pos_out, len, flags);
 }
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 3, 0)
 
 // no REMAP_FILE_DEDUP: use file_in
 // https://cs.android.com/android/kernel/superproject/+/common-android-mainline:common/fs/read_write.c;l=1598-1599;drc=398da7defe218d3e51b0f3bdff75147e28125b60
@@ -321,6 +324,7 @@
         return orig->f_op->remap_file_range(orig, pos_in, file_out, pos_out, len, remap_flags);
     }
 }
+#endif /* remap_file_range */
 
 static int ksu_wrapper_fadvise(struct file *fp, loff_t off1, loff_t off2, int flags)
 {
@@ -361,8 +365,9 @@
     p->ops.write = fp->f_op->write ? ksu_wrapper_write : NULL;
     p->ops.read_iter = fp->f_op->read_iter ? ksu_wrapper_read_iter : NULL;
     p->ops.write_iter = fp->f_op->write_iter ? ksu_wrapper_write_iter : NULL;
-    p->ops.iopoll = fp->f_op->iopoll ? ksu_wrapper_iopoll : NULL;
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 1, 0)
    p->ops.iopoll = fp->f_op->iopoll ? ksu_wrapper_iopoll : NULL;
 #if LINUX_VERSION_CODE < KERNEL_VERSION(6, 6, 0)
+#endif /* iopoll */
     p->ops.iterate = fp->f_op->iterate ? ksu_wrapper_iterate : NULL;
 #endif
     p->ops.iterate_shared = fp->f_op->iterate_shared ? ksu_wrapper_iterate_shared : NULL;
@@ -394,8 +399,9 @@
     p->ops.copy_file_range = fp->f_op->copy_file_range ? ksu_wrapper_copy_file_range : NULL;
     p->ops.remap_file_range = fp->f_op->remap_file_range ? ksu_wrapper_remap_file_range : NULL;
     p->ops.fadvise = fp->f_op->fadvise ? ksu_wrapper_fadvise : NULL;
-
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 3, 0)

 #if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 6, 0)
+#endif /* remap_file_range */
     p->ops.splice_eof = fp->f_op->splice_eof ? ksu_wrapper_splice_eof : NULL;
 #endif
 
